### ISSUE TITLE
Ability to start review as a bunch of comments

### ISSUE THREAD

**Issue by petRUShka** [harrisoncramer/gitlab.nvim#223]
## Feature Description

It is possible to add comments in `gitlab.nvim`.

When one uses the GitLab Web UI, it is possible to start a [review](https://docs.gitlab.com/ee/user/project/merge_requests/reviews/), add multiple comments to it, and then submit them all at once. It is also possible to add a meta-comment to the review itself.

It would be beneficial to have this functionality in the plugin.

**Comment by harrisoncramer** [comment]
This would be a good feature to have.

Gitlab has a different API for draft notes from normal notes, unfortunately, see: https://docs.gitlab.com/ee/api/draft_notes.html

It would therefore be a pretty substantial addition, since we'd have to write up CRUD operations for this resource, build them into the discussion tree and display them differently, and provide the ability to create/remove/edit them via the UI. Happy to work on it though.

**Comment by harrisoncramer** [comment]
Ah, sadly this is not yet supported in Go Gitlab which is the library we use when connecting to Gitlab. That library needs to build support before we can add it to this plugin. There is an open request here:

https://github.com/xanzy/go-gitlab/issues/1861

**Comment by harrisoncramer** [comment]
@petRUShka I've merged an MR into the Go Gitlab library that adds the APIs needed for this functionality and can work on this soon.

**Comment by petRUShka** [comment]
Great news!

**Comment by jakubbortlik** [comment]
Hi Harrison, do you think it would be possible to use the same API for creating/editing/removing draft comments and only decide that they will be added to a bulk review or applied ditectly with a different `perform_action` callback? This would be similar to the way comments are handled in gitlab online and it would not require users to create any extra keybindings. The problem I see is there currently only can be one action to perform (not considering linewise actions which are used for uploading files).

**Comment by harrisoncramer** [comment]
My initial thought was that the GET comments and notes APIs would be the same, the only difference would be that we would mark the "draft" comments differently in the tree. When you have any draft comments you'd have the ability to publish your review, which would have to be a new action.

Then when it comes to creating draft comments... we don't surface anything that provides you the option to add a draft comment right now, we just send them right away. To configure that, we would show it as another field under the create comment and create note popups, like Gitlab does in the browser GUI:

![Screenshot 2024-04-10 at 1 56 02 PM](https://github.com/harrisoncramer/gitlab.nvim/assets/32515581/096cd732-cc0a-480b-949c-56a7b783486a)

That data could then be passed to the `create_comment` function which would add a comment to the current review as a draft. It would also be configurable to be default true/false in the setup call. What do you think of this?

**Comment by jakubbortlik** [comment]
Here's a more concrete example of what I meant:
```diff
diff --git a/lua/gitlab/state.lua b/lua/gitlab/state.lua
index be8a026..202ce61 100644
--- a/lua/gitlab/state.lua
+++ b/lua/gitlab/state.lua
@@ -269,10 +269,14 @@ local function exit(popup, opts)
 end
 
 -- These keymaps are buffer specific and are set dynamically when popups mount
-M.set_popup_keymaps = function(popup, action, linewise_action, opts)
+M.set_popup_keymaps = function(popup, action_with_description, linewise_action, opts)
   if opts == nil then
     opts = {}
   end
+  -- If the `action_with_description` param was {action, "Description of action"}, then we could use
+  -- a more explicit description in vim.keymap.set, e.g.:
+  -- state.set_popup_keymaps(pipeline_popup, {M.retrigger, "Retrigger pipeline"}, M.see_logs)
+  local action, desc = unpack(action_with_description)
   if action ~= "Help" then -- Don't show help on the help popup
     vim.keymap.set("n", M.settings.help, function()
       local help = require("gitlab.actions.help")
@@ -283,13 +287,16 @@ M.set_popup_keymaps = function(popup, action, linewise_action, opts)
     vim.keymap.set("n", M.settings.popup.perform_action, function()
       local text = u.get_buffer_text(popup.bufnr)
       if opts.action_before_close then
-        action(text, popup.bufnr)
+        -- Here we could use the count to parametrize `action`, e.g., `vim.v.count = 0` means, no
+        -- count was given - this would be the default action, with `count > 0`, an alternate action
+        -- would be performed.
+        action(text, popup.bufnr, vim.v.count)
         vim.api.nvim_buf_delete(popup.bufnr, {})
       else
         vim.api.nvim_buf_delete(popup.bufnr, {})
-        action(text, popup.bufnr)
+        action(text, popup.bufnr, vim.v.count)
       end
-    end, { buffer = popup.bufnr, desc = "Perform action" })
+    end, { buffer = popup.bufnr, desc = desc})
   end
 
   if linewise_action ~= nil then
```

**Comment by harrisoncramer** [comment]
Hello, this feature is now in the develop branch. If you'd like to try it, then please check out that branch in your plugin configuration locally and rebuild the server.

I'm going to keep this in `develop` for probably some time before releasing to `main` since it represents a major overhaul of the plugin, and I'd like to avoid breaking existing functionality.

**Comment by jakubbortlik** [comment]
Hi Harrison, thanks for all the work on the draft functionality! Do you plan to enable editing draft notes? I believe this would be very useful. If this is not a priority at the moment, I'd at least suggest to make it impossible to actually edit draft notes. When I edit a draft note it still looks absolutely normal in `gitlab.nvim`, but on Gitlab, the draft content is lost (even if the draft is indicated in the list of drafts, see screenshot below), and when I publish the draft, it gets lost completely. Similarly, it is possible to start writing a reply to a draft note, but then an error is thrown when I want to publish the reply - it would be more user friendly to disable replying to drafts altogether.

Empty draft on Gitlab:
![image](https://github.com/harrisoncramer/gitlab.nvim/assets/17056143/789d7a41-7703-4d22-afbc-c861e6568879)

**Comment by harrisoncramer** [comment]
Hey Jakub, thank you for reviewing this. 

1. Do you plan to enable editing draft notes? 

This is supported, I was not sending the position data via the Lua API call, which is required. This is now fixed in develop. 

2. It is possible to start writing a reply to a draft note, but then an error is thrown when I want to publish the reply.

Gitlab does not support this. I've added a warning flash and blocked replying in the tree.

**Comment by harrisoncramer** [comment]
This is now supported on main, I've got replies cut separately because they're quite difficult https://github.com/harrisoncramer/gitlab.nvim/issues/286
